pandas/core/internals/methods.py

```diff
@@ -0,0 +1,114 @@
+"""
+Wrappers around array_algos with internals-specific logic
+"""
+from __future__ import annotations
+
+import numpy as np
+
+from pandas.core.dtypes.cast import (
+    can_hold_element,
+    find_common_type,
+    infer_dtype_from,
+)
+from pandas.core.dtypes.common import is_interval_dtype
+from pandas.core.dtypes.generic import (
+    ABCDataFrame,
+    ABCIndex,
+    ABCSeries,
+)
+from pandas.core.dtypes.missing import (
+    is_valid_na_for_dtype,
+    na_value_for_dtype,
+)
+
+from pandas.core.array_algos.putmask import (
+    extract_bool_array,
+    putmask_smart,
+    putmask_without_repeat,
+    setitem_datetimelike_compat,
+    validate_putmask,
+)
+from pandas.core.arrays import ExtensionArray
+from pandas.core.arrays._mixins import NDArrayBackedExtensionArray
+
+
+def putmask_flexible(array: np.ndarray | ExtensionArray, mask, new):
```

Comment:
where is all of this logic from? this doesn't look like 'simple' wrappers to me.

Reply:
It's logic that for the BlockManager lives on the blocks (the Block.putmask version also calls those `putmask_without_repeat` and `putmask_smart` helpers from `core.array_algos.putmask`

Reply:
great can u then remove from where it's used now

Reply:
The BlockManager is still used .. 
I know this gives some duplication, but the parts that can be shared are already in `pandas.core.array_algos.putmask`. Both the Block version as this version are using those functions. 

I can try to see if there are some more parts that could be moved into `array_algos.putmask`

Reply:
OK, so I could easily remove most of the logic of `ExtensionBlock.putmask` and reuse the new helper function. For base `Block.putmask` that will be more difficult, as that directly involves splitting blocks if needed etc (so Block-specific logic)

Reply:
let me put this another way. you are adding a ton of code to methods, is this generally useful? can we reuse elsewhere?

Reply:
> you are adding a ton of code to methods, is this generally useful? can we reuse elsewhere?

No, it is only for use within `array_manager.py`, but I put it in a separate `internals/methods.py` file to not make `array_manager.py` any longer (and on request of Brock I also didn't put it in `/core/array_algos/putmask.py` because this is the custom putmask logic for internals (with fallback logic etc that we need for Series/DataFrame), and not a general putmask algo)